File: Interview Problems/Parking Lot/Java Code/domain/PricingRule.java
package domain;

import java.util.UUID;

public class PricingRule {

    public enum RuleType {
        FLAT, HOURLY
    }

    private final UUID id;
    private final Vehicle.VehicleType vehicleType;
    private double ratePerHour;
    private double flatRate;
    private RuleType ruleType;

    public PricingRule(Vehicle.VehicleType vehicleType, double ratePerHour, double flatRate, RuleType ruleType) {
        this.id = UUID.randomUUID();
        this.vehicleType = vehicleType;
        this.ratePerHour = ratePerHour;
        this.flatRate = flatRate;
        this.ruleType = ruleType;
    }

    public void updateFlatRate(double flatRate) { this.flatRate = flatRate; }
    public void updateHourlyRate(double ratePerHour) { this.ratePerHour = ratePerHour; }
    public void updateRuleType(RuleType ruleType) { this.ruleType = ruleType; }

    public UUID getId() { return id; }
    public Vehicle.VehicleType getVehicleType() { return vehicleType; }
    public double getRatePerHour() { return ratePerHour; }
    public double getFlatRate() { return flatRate; }
    public RuleType getRuleType() { return ruleType; }

    @Override
    public String toString() {
        return "PricingRule{vehicleType=" + vehicleType + ", flatRate=" + flatRate
                + ", ratePerHour=" + ratePerHour + ", ruleType=" + ruleType + "}";
    }
}

File: Interview Problems/Parking Lot/Python Code/domain/pricing_rule.py
import uuid
from enum import Enum, auto
from domain.vehicle import VehicleType

class RuleType(Enum):
    FLAT = auto()
    HOURLY = auto()

class PricingRule:
    def __init__(self, vehicle_type: VehicleType, rate_per_hour: float, flat_rate: float, rule_type: RuleType):
        self.id = uuid.uuid4()
        self.vehicle_type = vehicle_type
        self.rate_per_hour = rate_per_hour
        self.flat_rate = flat_rate
        self.rule_type = rule_type

    def update_flat_rate(self, flat_rate: float):
        self.flat_rate = flat_rate

    def update_hourly_rate(self, rate_per_hour: float):
        self.rate_per_hour = rate_per_hour

    def update_rule_type(self, rule_type: RuleType):
        self.rule_type = rule_type

    def __repr__(self):
        return f"PricingRule(vehicle_type={self.vehicle_type.name}, flat_rate={self.flat_rate}, rate_per_hour={self.rate_per_hour}, rule_type={self.rule_type.name})"
